From dc53c5378c868b50ff90528955d63f9397b634b6 Mon Sep 17 00:00:00 2001
From: Aymen Sghaier <aymen.sghaier@nxp.com>
Date: Mon, 4 Jun 2018 11:21:28 +0200
Subject: [PATCH 4197/5242] MLK-17304: crypto: caam: Fix error swiotlb buffer
 is full for caamhash

commit  3d150b37cf049a1df1573cb61b2d69b9e1a56551 from
https://source.codeaurora.org/external/imx/linux-imx.git

 During caamhash tests the error "swiotlb buffer is full" occurred.
 This was due to dma mapping without unmapping later.
 This patch adds the unmap call to avoid the loss of dma memory.

Signed-off-by: Aymen Sghaier <aymen.sghaier@nxp.com>
Signed-off-by: Franck LENORMAND <franck.lenormand@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/crypto/caam/caamhash.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/crypto/caam/caamhash.c b/drivers/crypto/caam/caamhash.c
index 2f81663..3cda0bf 100644
--- a/drivers/crypto/caam/caamhash.c
+++ b/drivers/crypto/caam/caamhash.c
@@ -730,6 +730,14 @@ static inline void ahash_unmap(struct device *dev,
 	if (edesc->dst_dma)
 		dma_unmap_single(dev, edesc->dst_dma, dst_len, DMA_FROM_DEVICE);
 
+	if (state->buf_dma) {
+		dma_unmap_single(dev, state->buf_dma,
+			(state->current_buf ?
+				state->buflen_1 : state->buflen_0),
+			DMA_TO_DEVICE);
+		state->buf_dma = 0;
+	}
+
 	if (edesc->sec4_sg_bytes)
 		dma_unmap_single(dev, edesc->sec4_sg_dma,
 				 edesc->sec4_sg_bytes, DMA_TO_DEVICE);
-- 
1.7.9.5

